dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java

```diff
@@ -511,55 +619,183 @@ private WorkflowVersion setupWorkflowFilesForVersion(String calculatedPath, Trip
+            String msg = "Invalid .dockstore.yml";
+            LOG.info(msg, ex);
+            return null;
+        }
+
+        String primaryDescriptorPath = (String)workflowMap.get("primaryDescriptorPath");
+        if (primaryDescriptorPath == null)  {
+            String msg = ".dockstore.yml is missing the required primaryDescriptorPath field.";
+            LOG.info(msg);
+            return null;
+        }
+        String fileContent = this.readFileFromRepo(primaryDescriptorPath, ref.getLeft(), repository);
+        if (fileContent != null) {
+            // Add primary descriptor file and resolve imports
+            SourceFile file = new SourceFile();
+            file.setAbsolutePath(primaryDescriptorPath);
+            file.setPath(primaryDescriptorPath);
+            file.setContent(fileContent);
+            DescriptorLanguage.FileType identifiedType;
+            if (Objects.equals(workflow.getDescriptorType(), DescriptorLanguage.CWL)) {
+                identifiedType = DescriptorLanguage.FileType.DOCKSTORE_CWL;
+            } else if (Objects.equals(workflow.getDescriptorType(), DescriptorLanguage.WDL)) {
+                identifiedType = DescriptorLanguage.FileType.DOCKSTORE_WDL;
+            } else if (Objects.equals(workflow.getDescriptorType(), DescriptorLanguage.NEXTFLOW)) {
+                identifiedType = DescriptorLanguage.FileType.NEXTFLOW_CONFIG;
+            } else {
+                String msg = "Invalid type";
+                LOG.info(msg);
+                return null;
             }
-            for (String filePath: files) {
-                String fileContent = this.readFileFromRepo(filePath, ref.getLeft(), repository);
-                if (fileContent != null) {
-                    SourceFile file = new SourceFile();
-                    file.setAbsolutePath(filePath);
-                    file.setPath(filePath);
-                    file.setContent(fileContent);
-                    file.setType(DescriptorLanguage.FileType.DOCKSTORE_SERVICE_OTHER);
-                    version.getSourceFiles().add(file);
-                } else {
-                    // File not found or null
-                    LOG.info("Could not find file " + filePath + " in repo " + repository);
+            file.setType(identifiedType);
+            version.setWorkflowPath(primaryDescriptorPath);
+
+            version = combineVersionAndSourcefile(repository.getFullName(), file, workflow, identifiedType, version, existingDefaults);
+
+            if (testParameterPaths != null) {
+                for (String testParameterPath : testParameterPaths) {
+                    String testJsonContent = this.readFileFromRepo(testParameterPath, ref.getLeft(), repository);
+                    if (testJsonContent != null) {
+                        SourceFile testJson = new SourceFile();
+
+                        if (identifiedType.equals(DescriptorLanguage.FileType.DOCKSTORE_CWL)) {
```

Comment:
This can use https://github.com/dockstore/dockstore/pull/3235/files#diff-fd5ac5b6f6fdba95c1c5be457e049d7dR207 after my merge